src/tool/xmeta/cpp/models/namespace_model.cpp

```diff
@@ -0,0 +1,130 @@
+#include "namespace_model.h"
+
+#include "using_directive_models.h"
+
+#include "class_model.h"
+#include "delegate_model.h"
+#include "enum_model.h"
+#include "interface_model.h"
+#include "property_model.h"
+#include "struct_model.h"
+
+namespace xlang::xmeta
+{
+    void namespace_body_model::add_class(std::shared_ptr<class_model> const& cm)
+    {
+        m_classes[cm->get_id()] = cm;
+    }
+
+    void namespace_body_model::add_struct(std::shared_ptr<struct_model> const& sm)
+    {
+        m_structs[sm->get_id()] = sm;
+    }
+
+    void namespace_body_model::add_interface(std::shared_ptr<interface_model> const& im)
+    {
+        m_interfaces[im->get_id()] = im;
+    }
+
+    void namespace_body_model::add_enum(std::shared_ptr<enum_model> const& em)
+    {
+        m_enums[em->get_id()] = em;
+    }
+
+    void namespace_body_model::add_delegate(std::shared_ptr<delegate_model> const& dm)
+    {
+        m_delegates[dm->get_id()] = dm;
+    }
+
+    void namespace_body_model::add_using_alias_directive(std::shared_ptr<using_alias_directive_model> const& uad)
+    {
+        m_using_alias_directives[uad->get_id()] = uad;
+    }
+
+    void namespace_body_model::add_using_namespace_directive(std::shared_ptr<using_namespace_directive_model> const& und)
+    {
+        m_using_namespace_directives.emplace_back(std::move(und));
+    }
+
+    std::string copy_to_lower(std::string_view sv)
+    {
+        std::string s{ sv };
+        std::locale loc;
+        // ***NOTE*** ::tolower is temporary. This only works on the ASCII charset.
+        // Comparing case in full UTF-8 is a larger discussion, and should first be documented in detail.
+        std::transform(s.begin(), s.end(), s.begin(), [&](char c)
+        {
+            return static_cast<char>(::tolower(c));
+        });
+        return s;
+    }
+
+    // Checks if two member IDs are semantically the same.
+    // namespace member IDs are case preserving but insensitive. This means if 'classA' is defined,
+    // you cannot define another member 'cLaSsA'. You also cannot reference 'classA' using 'cLaSsA'
+    // since case is preserved.
+    template<typename T>
```

Comment:
How often will this check happen? It's creating two new strings every time it's called.

Reply:
We can use stricmp from string.h, but it's only Windows. Posix-based machines have strcasecmp from strings.h. I'll update the issue, since we're not using ::tolower anymore.